OpenSim/Actuators/VandenBogert2011Muscle.cpp

```diff
@@ -0,0 +1,1234 @@
+const
+//TODO: Add clamping?  Add ignore_activation_dynamics
+{
+    setStateVariableValue(s, "activation", activation);
+
+    //markCacheVariableInvalid(s,"velInfo");
+    //markCacheVariableInvalid(s,"dynamicsInfo");
+}
+
+// Get the Residual of muscle
+// TODO:  This breaks naming convention in that it does not get an already
+// evaluated constant, it actually performs the calculation.  Without using a
+// state variable or adding to the cache variables, I think I have to do it this
+// way?
+SimTK::Vec2 VandenBogert2011Muscle::getResidual(const SimTK::State& s,
+                                                double projFibVelNorm_guess,
+                                                double activdot_guess, double u)
+                                                const
+{
+    ImplicitResidual residualStruct = calcImplicitResidual(s,
+                                                           projFibVelNorm_guess,
+                                                           activdot_guess,
+                                                           u, 0);
+
+    SimTK::Vec2 residual = {residualStruct.forceResidual,
+                            residualStruct.activResidual};
+    return(residual);
+}
+
+
+
+//These are hacks because I am not using muscle.h mli cache variables, yet:
+double  VandenBogert2011Muscle::getFiberLength(SimTK::State& s) const
+{
+    double fiberLength=getStateVariableValue(s, "fiber_length");
+    return fiberLength;
+}
+double  VandenBogert2011Muscle::getActivation(SimTK::State& s) const
+{
+    double activ=getStateVariableValue(s, "activation");
+    return activ;
+}
+
+
+
+//==============================================================================
+// Muscle.h Interface
+//==============================================================================
+
+
+double  VandenBogert2011Muscle::computeActuation(const SimTK::State& s) const
+{
+    double activ=getActivation(s);
+    // For now we will start with static guess
+    // TODO: figure out how to use previous state as guess
+    SimTK::Vector ydotInitialGuess(2);
+    ydotInitialGuess[0]=0.0;
+    ydotInitialGuess[1]=0.0;
+    SimTK::Vec2 sdot=calcSolveMuscle(s, activ, ydotInitialGuess);
+
+    VandenBogert2011Muscle::ImplicitResidual Result= calcImplicitResidual(s,
+                                          sdot[0],
+                                          sdot[1],
+                                          activ,
+                                          0);
+
+    //TODO: Could probably add a check here that residuals =0
+    double actuation=Result.forceTendon;
+
+    return(actuation);
+}
+
+
+
+void VandenBogert2011Muscle::computeInitialFiberEquilibrium(SimTK::State& s)
+const
+{
+    // TODO:  this is the same code as computeFiberEquilibriumAtZeroVelocity
+    //Need to update for v~=0
+    //Let's start with the default activation and the fiberVelocity=0
+
+    computeFiberEquilibriumAtZeroVelocity(s);
+
```

Comment:
> I think this is needlessly complex for a constraint that only needs to be approximate.

Probably. A related thought: it might be possible to simplify the algorithm used in the Millard muscle by splitting the problem into three parts. Because of the shape of the (active+passive) force–length curve, between one and three fiber lengths correspond to a given force; Millard's algorithm doesn't take advantage of this fact and instead solves a more general problem (potentially slower and less reliably). It might be possible to split the search for the fiber length into three pieces:

![forcelengthcurve](https://cloud.githubusercontent.com/assets/4203505/17605473/6cb5148a-5fcf-11e6-8fda-0de998afddd8.png)

Each piece is a well-behaved curve where searching should be very fast even with only an approximate derivative.